### ISSUE TITLE
[Llama-3.1-8B] Optimize ccls in `tt_transformers` to match/surpass old performance

### ISSUE THREAD

**Issue by rdraskicTT** [tenstorrent/tt-metal#33394]
[Migration to new ccls](https://github.com/tenstorrent/tt-metal/pull/25628) caused a degradation in decode and prefill times. Here are the results on batch=1, T3K submesh on galaxy for Llama 8B in TP=8.

| Phase   | Before (t/s) | After (t/s) | Drop (%) |
|---------|-------------|------------|-----------------|
| Decode  | 66          | 55         | 17%           |

**Comment by rdraskicTT** [comment]
Profile for a single decode layer on commit before the fabric change

<img width="2394" height="1038" alt="Image" src="https://github.com/user-attachments/assets/6c48f642-7890-4544-8ade-fa094d56a7fc" />

Profile for a single decode layer on latest main

<img width="2284" height="1010" alt="Image" src="https://github.com/user-attachments/assets/af35cdfe-e49e-42d2-999a-b252434f7c2d" />

We see that ccls take longer and use less cores, which should be investigated.

**Comment by mstojkoTT** [comment]
Use the branch `mstojko/regression_test`. Run by modyfing `.vscode/launch.json` so the file has this:
```
    "configurations": [
        {
            "name": "Py Dbg",
            "type": "python",
            "python": "${workspaceFolder}/python_env/bin/python",
            "request": "launch",
            "module": "tracy",
            "console": "integratedTerminal",
            "justMyCode": true,
            "env": {

                "HF_MODEL": "meta-llama/Llama-3.1-8B-Instruct",
                "MESH_DEVICE": "TG",
            },
        "args": [

                "-r", "-v", "-p", "-m", "pytest",

                "models/tt_transformers/tests/test_decoder.py",
                "-x"

            ],
            "stopOnEntry": false
        },
```
For an end to end test you can run `models/tt_transformers/demo/simple_text_demo.py -k performance and batch-1`. For a more "unit test"-ish situation you can `run models/tt_transformers/tests/test_decoder.py` with tracy (via the below defined `launch.json`, and see how it has regressed for the CCLs (this can be seen in the above images @rdraskicTT posted). You can see the regression by comparing the current commit on main with the old one before the PR was merged.

**Comment by mstojkoTT** [comment]
This is the tracy run for `models/tt_transformers/tests/test_decoder.py`. For example Reduce Scatter seems a lot slower
[old_before_pr.csv](https://github.com/user-attachments/files/23858361/old_before_pr.csv)
[new_main.csv](https://github.com/user-attachments/files/23858362/new_main.csv)

**Comment by sjameelTT** [comment]
Can I get a CLI command for the decoder test? As well, could you tell me the file/line number of the CCL calls so I can just go straight there.

**Comment by mstojkoTT** [comment]
> Can I get a CLI command for the decoder test? As well, could you tell me the file/line number of the CCL calls so I can just go straight there.

`models/tt_transformers/tt/mlp.py` line 257 commit `99846a4b53`. 


As for the CLI, does the launch.json work for you or do you want a command?

**Comment by sjameelTT** [comment]
So far, for Galaxy we can look to special case by using 4 links rather than the 1 for T3K. Let's try this experiment and see our numbers.

**Comment by ekuslysTT** [comment]
@ppetrovicTT WIP

**Comment by ekuslysTT** [comment]
Notes
- @cglagovich @jonathansuTT to lean in. sync with @ppetrovicTT .
- @ppetrovicTT update action plan/next steps

**Comment by zzigler-tt** [comment]
@ppetrovicTT can you please provide daily updates on the status, next steps, etc. Thanks

**Comment by ekuslysTT** [comment]
Notes
- CCLs on TTT are in bad shape. None of the glx optimizations are used. Even though they are being called.
- @ppetrovicTT working on list of to dos, working with @jonathansuTT  and @cglagovich .
- @ppetrovicTT also include @cmaryanTT @ntarafdar @sjameelTT @jvegaTT

**Comment by ekuslysTT** [comment]
Notes
- focusing on CCLs with CCLs team to find best combination of parameters

**Comment by ekuslysTT** [comment]
Notes
- slow progress, low confidence to fix this week. Long term improvement

**Comment by jonathansuTT** [comment]
<img width="1179" height="532" alt="Image" src="https://github.com/user-attachments/assets/e0e97077-a1cc-4a6b-8a51-86846210ab8f" />

The total with legacy CCLs was 413us, after was 474us.  With the following changes, it is down to 443us:
mlp.py:269: num_reduce_scatter_links=4,#self.args.num_reduce_scatter_links

The slow down comes from 2 places:
on FF1, FF2, and FF3, the matmul is 40/40/44us instead of 34/34/34us.  Other 1D matmuls earlier are fine, so most likely matmul1D config has changed slightly.
TODO: revert to legacy CCL commit, run, check what the configs are for FF1,FF2, and FF3.

ReduceScatter is still suboptimal.  This is a small shape, so more links doesn't really help too much (at least with the current code).  Even so, we have shaved down from 67 to 55us by increasing num_links from 1 to 4.  Adding more workers per link did not help.
TODO: Switch to fused MatmulReduceScatter, this will prob get us closer to 44us. 

The 3 matmuls account for 20us of regression, and the reduce scatter accounts for 10us.

**Comment by ppetrovicTT** [comment]
**E2E measurements, Galaxy wh-glx6u-02, DP=4**

**TL;DR**
The target is 67 t/s/u
- Current
  - simple_text_demo measures 34.2 t/s/u
  - vLLM benchmark measures 49.18 t/s/u
- Pre-fabric
  - simple_text_demo measures  52.89 t/s/u
  - vLLM benchmark measures 48.78 t/s/u

----------------------------------------------------

**Current state**
- tt-metal commit: 0a46baeb70753aee85cb83b336df9968c27e812d
- vllm commit: [b2894d34fed97137713b845051717de56bc129cf](https://github.com/tenstorrent/vllm/commit/b2894d34fed97137713b845051717de56bc129cf)

simple_text_demo run:
```
#!/bin/bash

export TT_METAL_HOME="/localdev/ppetrovic/tt-metal"
export PYTHON_ENV_DIR="$TT_METAL_HOME/python_env"
export PYTHONPATH=$TT_METAL_HOME

export MESH_DEVICE="TG"
export HF_MODEL="meta-llama/Llama-3.1-8B-Instruct"

export TT_MM_THROTTLE_PERF=5

TEST=(models/tt_transformers/demo/simple_text_demo.py -k "performance-DP-4-b32")

python -m pytest "${TEST[@]}"
```

simple_text_demo result:
```
Average Time to First Token (TTFT): 10.57ms
Average speed: 29.24ms @ 34.2 tok/s/user (4377.62 tok/s throughput)
```

vllm run:
```
### Start server
#!/bin/bash

# Set environment variables
export MESH_DEVICE="TG"
export TT_METAL_HOME="/localdev/ppetrovic/tt-metal"
export PYTHON_ENV_DIR="$TT_METAL_HOME/python_env"
export PYTHONPATH=$TT_METAL_HOME

export vllm_dir="/localdev/ppetrovic/vllm"
export VLLM_RPC_TIMEOUT="900000"
export VLLM_TARGET_DEVICE="tt"
export VLLM_PORT="8000"

export TT_MM_THROTTLE_PERF=5

export HF_MODEL="meta-llama/Llama-3.1-8B-Instruct"
export TT_CACHE_PATH="/localdev/ppetrovic/hf_data/tt_cache/meta-llama/Llama-3.1-8B-Instruct"

# Run the server
python $vllm_dir/examples/server_example_tt.py \
    --model $HF_MODEL \
    --block_size 64 \
    --max_model_len 65536 \
    --max_num_seqs 128 \
    --max_num_batched_tokens 65536 \
    --num_scheduler_steps 10 \
    --override_tt_config '{"trace_region_size": 50000000, "data_parallel": 4, "sample_on_device_mode": "decode_only"}' \
    --max-log-len 32 \
    --port $VLLM_PORT \
    2>&1 | tee "vllm_server_$(date +%Y%m%d_%H%M%S).log"
```
```
### Start benchmark
FILE_BENCHMARK_LOG="vllm_benchmark.log"
MODEL="meta-llama/Llama-3.1-8B-Instruct"
NUM_PROMPTS=256
vllm bench serve \
    --model $MODEL \
    --num-prompts $NUM_PROMPTS \
    --dataset-name random \
    --random-input-len 100 \
    --random-output-len 100 \
    --ignore-eos \
    --percentile-metrics ttft,tpot,itl,e2el \
    --save-result \
    --max-concurrency 32 \
    2>&1 | tee ${FILE_BENCHMARK_LOG}
```

vllm results:
```
============ Serving Benchmark Result ============
Successful requests:                     256       
Benchmark duration (s):                  24.92     
Total input tokens:                      25269     
Total generated tokens:                  25600     
Request throughput (req/s):              10.27     
Output token throughput (tok/s):         1027.20   
Total Token throughput (tok/s):          2041.11   
---------------Time to First Token----------------
Mean TTFT (ms):                          1129.88   
Median TTFT (ms):                        1127.48   
P99 TTFT (ms):                           1152.58   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          20.02     
Median TPOT (ms):                        19.85     
P99 TPOT (ms):                           20.88     
---------------Inter-token Latency----------------
Mean ITL (ms):                           20.02     
Median ITL (ms):                         18.30     
P99 ITL (ms):                            34.29     
----------------End-to-end Latency----------------
Mean E2EL (ms):                          3111.56   
Median E2EL (ms):                        3109.16   
P99 E2EL (ms):                           3191.34 
```
These results match the run from Models CI
https://github.com/tenstorrent/tt-shield/actions/runs/20248950648/job/58138883985#step:11:38

----------------------------------------------------

**Pre-fabric state**
- tt-metal commit: 6f16f486a5e49e5e4c86dd5225eac135b7f3b458
- vllm commit: [128616e48cb9e8cdd04b14d416f7ae338beb7a42](https://github.com/tenstorrent/vllm/commit/128616e48cb9e8cdd04b14d416f7ae338beb7a42)

simple_text_demo run - same as above.
simple_text_demo result:
```
Average Time to First Token (TTFT): 90.12ms
Average speed: 18.91ms @ 52.89 tok/s/user (6770.29 tok/s throughput)
```
Much higher than 34.2 t/s/u from above, but matching with vLLM benchmark, and still far from 67 t/s/u target.

vllm run - server same as above, benchmark:
```
FILE_BENCHMARK_LOG="vllm_benchmark.log"
MODEL="meta-llama/Llama-3.1-8B-Instruct"
NUM_PROMPTS=256
vllm_dir="/localdev/ppetrovic/vllm"
python $vllm_dir/benchmarks/benchmark_serving.py \
    --model $MODEL \
    --backend vllm \
    --endpoint /v1/completions \
    --port 8000 \
    --num-prompts $NUM_PROMPTS \
    --dataset-name random \
    --random-input-len 100 \
    --random-output-len 100 \
    --ignore-eos \
    --percentile-metrics ttft,tpot,itl,e2el \
    --save-result \
    --max-concurrency 32 \
    2>&1 | tee ${FILE_BENCHMARK_LOG}
```

vllm results:
```
============ Serving Benchmark Result ============
Successful requests:                     256       
Benchmark duration (s):                  39.58     
Total input tokens:                      25600     
Total generated tokens:                  25600     
Request throughput (req/s):              6.47      
Output token throughput (tok/s):         646.82    
Total Token throughput (tok/s):          1293.64   
---------------Time to First Token----------------
Mean TTFT (ms):                          2914.48   
Median TTFT (ms):                        2838.20   
P99 TTFT (ms):                           3385.63   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          20.50     
Median TPOT (ms):                        19.87     
P99 TPOT (ms):                           22.73     
---------------Inter-token Latency----------------
Mean ITL (ms):                           20.50     
Median ITL (ms):                         19.26     
P99 ITL (ms):                            35.78     
----------------End-to-end Latency----------------
Mean E2EL (ms):                          4943.65   
Median E2EL (ms):                        4845.48   
P99 E2EL (ms):                           5491.16   
==================================================
```

**Comment by jonathansuTT** [comment]
We have found the cause of the matmul slowdown (20us)
https://github.com/tenstorrent/tt-metal/commit/7bbf93b339a5baadabb551a826ce6f09d425dd7a
Changing core grid to 7,7 instead of 7,8 fixes all the matmul issues.  We will meet with Almeet this morning to come up with a more elegant commit-able solution.

The only remaining delta (10us out of the original 61us) is reduce scatter, which should be resolved with fusing, but I'll look at that later today (though we should already be within acceptable time).  

We also need to get all these changes on main.

**Comment by jonathansuTT** [comment]
We fixed the matmul issues with a core grid override.
We optimized the CCL params further.
As all CCL's now use 1 worker per link (since they're all 32x tensors), we then removed mux cores.
With this, we are down to 406us per transformer block, which is 7us faster than before legacy ccl's were removed.
<img width="1253" height="492" alt="Image" src="https://github.com/user-attachments/assets/0fba6afc-12e4-45ed-b528-aa23e0ab2c03" />

2 more items:
1) need to figure out to how merge non-mux ccl's gracefully, but this should be easy/low-risk.
2) need to remove overhead outside of transformer block. @mbahnasTT is looking at that

**Comment by ekuslysTT** [comment]
Notes
Still need to run E2E via simple text demo --> machine availability

**Comment by jonathansuTT** [comment]
had to add in a switch between mux and non-mux versions of the CCL (@aperezvicente-TT and @alnah005 did this), so that prefill and decode could run back to back to run the e2e test.

the measured per transformer block decode time is down to 402us, but there is some overhead still in e2e so it's still at 56 T/S/U, looking at the non-transformer block overhead now

**Comment by jonathansuTT** [comment]
rewriting sampling to replace sampling with just argmax in tt_sampling, this should save about 4ms

**Comment by ekuslysTT** [comment]
@jonathansuTT can you please re-post your comment? It's not showing up. Thanks!

**Comment by jonathansuTT** [comment]
tsu perf is ok now, 64.5 ts on batch 32.  the ci accuracy tests are failing, will investigate this today.  api tests are also failing (this worries me less, this is prob just a result of the sampling API changes we put in, we just need to reconcile this against all the tests).

**Comment by jonathansuTT** [comment]
thanks @handrewsTT and @sraizada-tt for the help debugging determinism test errors.  there are still 2 more outstanding issues.  non-uniform seed api test still fails, and accuracy test fails when trace is not True.  debugging those

**Comment by jonathansuTT** [comment]
just non-uniform seed api test failing left

**Comment by rdraskicTT** [comment]
@jonathansuTT Can you check if you are seeing the same thing we are seeing on llama 70b seed issue https://github.com/tenstorrent/tt-metal/issues/34610
We see batched prefill producing different logits per user for same inputs and also seed getting corrupted on some cores.

**Comment by jonathansuTT** [comment]
all issues should be fixed now

**Comment by jonathansuTT** [comment]
> [@jonathansuTT](https://github.com/jonathansuTT) Can you check if you are seeing the same thing we are seeing on llama 70b seed issue [#34610](https://github.com/tenstorrent/tt-metal/issues/34610) We see batched prefill producing different logits per user for same inputs and also seed getting corrupted on some cores.

no it is unrelated to what you guys are seeing in #34610

**Comment by bgoelTT** [comment]
Next steps:
1. Qwen3-32B & Llama-3.3-70B-Instruct regressions fixed on branch
2. Llama-3.1-8B-Instruct was raising an exception in Models CI
  a. needs rebase to latest commits on `stable` to accommodate prefill sampling

**Comment by bgoelTT** [comment]
Removing this from C1 tracking as target met. This work needs to be upstreamed into tt-metal `main` so leaving the issue open

**Comment by zzigler-tt** [comment]
@bgoelTT if we are removing this issue from C1 should it also be downgraded to priority P1 as this is not an urgent item to remediate since targets were met?

**Comment by bgoelTT** [comment]
It should still be a P0 tt-metal priority because we know how to achieve the perf, and we need this upstreamed to `main` for future releases. We won't always be working on this `stable` release branch
